File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int res=0;
        int lm=0;
        int rm=0;
        int l=0;
        int n = height.length;
        int r=n-1;
        while(l<r){
            if(height[l]<=height[r]){
                if(height[l]>lm){
                    lm=height[l];
                }
                else{
                    res+=lm-height[l];
                }
                l++;
            }
            else{
                if(height[r]>rm){
                    rm=height[r];
                }
                else{
                    res+=rm-height[r];
                }
                r--;
            }
        }
        return res;
    }
}

File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n=height.size();
        vector<int> lmax(n);
        int max=height[0];
        for(int i=0;i<n;i++){
            if(height[i]>max) max=height[i];
            lmax[i]=max;
        }
        max = height[n-1];
        vector<int> rmax(n);
        for(int i=n-1;i>=0;i--){
            if(height[i]>max) max=height[i];
            rmax[i]=max;
        }
        int sum=0;
        for(int i=0;i<n;i++){
            int min1=min(lmax[i],rmax[i]);
            sum+=min1-height[i];
        }
        return sum;
    }
};